drivers/comparator/comparator_it51xxx_vcmp.c

```diff
@@ -0,0 +1,451 @@
+				sys_write8(sys_read8(reg_base + vcmp_ctrl_reg[idx]) &
+						   ~VCMP_GREATER_THRESHOLD,
+					   reg_base + vcmp_ctrl_reg[idx]);
+			} else {
+				sys_write8(sys_read8(reg_base + vcmp_ctrl_reg[idx]) |
+						   VCMP_GREATER_THRESHOLD,
+					   reg_base + vcmp_ctrl_reg[idx]);
+			}
+
+			/* W/C voltage comparator specific channel interrupt status */
+			clear_vcmp_status(dev, idx);
+		}
+	}
+
+	/* W/C voltage comparator irq interrupt status */
+	ite_intc_isr_clear(cfg->irq);
+}
+
+static int vcmp_it51xxx_init(const struct device *dev)
+{
+	const struct vcmp_it51xxx_config *const cfg = dev->config;
+	struct vcmp_it51xxx_data *const data = dev->data;
+	uintptr_t base_ch = cfg->base_ch;
+	uintptr_t reg_vcmpscp2 = cfg->reg_base + REG_VCMP_VCMPSCP2;
+	uint8_t reg_value;
+	int err;
+
+	data->adc_ch_cfg.gain = ADC_GAIN_1;
+	data->adc_ch_cfg.reference = ADC_REF_INTERNAL;
+	data->adc_ch_cfg.acquisition_time = ADC_ACQ_TIME_DEFAULT;
+	data->adc_ch_cfg.channel_id = cfg->channel_id;
+
+	/* Disable voltage comparator and interrupt */
+	vcmp_enable(dev, false);
+
+	/*
+	 * ADC channel signal output to voltage comparator,
+	 * so we need to set ADC channel to alternate mode first.
+	 */
+	if (!device_is_ready(cfg->adc)) {
+		LOG_ERR("ADC device not ready");
+		return -ENODEV;
+	}
+
+	err = adc_channel_setup(cfg->adc, &data->adc_ch_cfg);
+	if (err) {
+		return err;
+	}
+
+	/* Select which ADC channel output voltage into comparator */
+	reg_value = FIELD_PREP(GENMASK(7, 3), sys_read8(base_ch));
+	reg_value |= data->adc_ch_cfg.channel_id & VCMP_CHANNEL_ID_REG_MASK;
+	sys_write8(reg_value, base_ch);
+
+	/* Set VCMP to Edge Sense Mode */
+	sys_write8(sys_read8(base_ch) | VCMP_CMP_EDGE_SENSED_MODE, base_ch);
+
+	if (cfg->vcmp_ch >= VCMP_CHANNEL_CNT) {
+		LOG_ERR("invalid volt comparator channel setting(%d)", cfg->vcmp_ch);
+		return -EINVAL;
+	}
+	vcmp_data[cfg->vcmp_ch] = dev->data;
+
+	/*
+	 * Set minimum scan period for "all" voltage comparator
+	 * Three voltage comparators share a scan period setting and use the fastest one
+	 */
+	if (cfg->scan_period < FIELD_GET(SCAN_PERIOD_MASK, sys_read8(reg_vcmpscp2))) {
+		sys_write8(FIELD_PREP(SCAN_PERIOD_MASK, cfg->scan_period), reg_vcmpscp2);
+	}
+
+	/* Data must keep device reference for worker handler */
+	data->vcmp = dev;
+
+	/* Init and set work item to enable notifications */
+	k_work_init(&data->work, it51xxx_vcmp_trigger_work_handler);
+	vcmp_work_addr[cfg->vcmp_ch] = (uint32_t)&data->work;
+
+	err = vcmp_set_threshold(dev, cfg->threshold_mv);
+	if (err) {
+		return err;
+	}
+
+	/*
+	 * All voltage comparator channels share one irq interrupt,
+	 * so if the irq is enabled before, we needn't to enable again.
+	 * And we will figure out the triggered channel in vcmp_it51xxx_isr().
+	 */
+	if (!irq_is_enabled(cfg->irq)) {
+		ite_intc_isr_clear(cfg->irq);
+		irq_connect_dynamic(cfg->irq, 0, (void (*)(const void *))vcmp_it51xxx_isr,
+				    (const void *)dev, 0);
+		irq_enable(cfg->irq);
+	}
```

Comment:
Will this not result in only one irq handler for one comparator be connected? `irq_connect_dynamic` must be called for all of them right? with only one irq_enable(cfg->irq)?

In any case, calling irq_enable(cfg->irq) multiple times is safe, so for simplicity I would just call it every time :)